docs/spec/dataclasses.rst

```diff
@@ -526,6 +532,56 @@ This includes, but is not limited to, the following semantics:
   dataclass must be explicitly annotated as e.g. ``x: ClassVar[Final[int]] =
   3``.
 
+Converters
+^^^^^^^^^^
+
+The ``converter`` parameter can be specified in a field definition to provide a single-argument callable
+used to convert values when assigning to the associated attribute. This feature allows for automatic type
+conversion and validation during object initialization and attribute assignment.
+
+Converter behavior:
+
+* For non-frozen dataclasses, the converter is used for all attribute assignment, including assignment
+  of default values and direct attribute setting (e.g., ``obj.attr = value``).
+* For frozen dataclasses, the converter is only used inside a ``dataclass``-synthesized ``__init__``
+  when setting the attribute.
+* The converter is not used when reading attributes, as the attributes should already have been converted.
+
+Typing rules for converters:
+
+* The ``converter`` must be a callable that accepts a single positional argument.
+* The parameter type of this single argument provides the type of the synthesized ``__init__`` parameter
+  associated with the field.
+* The return type of the callable must be assignable to the field's declared type.
+
+When used with ``default`` or ``default_factory``:
+
+* If a ``converter`` is provided alongside ``default`` or ``default_factory``, the type of the default
+  value should be assignable to the type of the single argument to the ``converter`` callable.
+* Default values are unconditionally converted using the ``converter`` if provided.
```

Comment:
I recommend moving this point up to the first sentence, as in:

"If a ``converter`` is provided alongside ``default`` or ``default_factory``, default values are converted using the ``converter``".

Then you can explain how a type checker should validate this behavior.